README.md
RAUC hawkBit Updater
====================

[![Build Status](https://github.com/rauc/rauc-hawkbit-updater/workflows/tests/badge.svg)](https://github.com/rauc/rauc-hawkbit-updater/actions)
[![License](https://img.shields.io/badge/license-LGPLv2.1-blue.svg)](https://raw.githubusercontent.com/rauc/rauc-hawkbit-updater/master/LICENSE)
[![Total alerts](https://img.shields.io/lgtm/alerts/g/rauc/rauc-hawkbit-updater.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/rauc/rauc-hawkbit-updater/alerts/)
[![Language grade: C/C++](https://img.shields.io/lgtm/grade/cpp/g/rauc/rauc-hawkbit-updater.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/rauc/rauc-hawkbit-updater/context:cpp)
[![Documentation](https://readthedocs.org/projects/rauc-hawkbit-updater/badge/?version=latest)](https://rauc-hawkbit-updater.readthedocs.io/en/latest/?badge=latest)
[![Matrix](https://img.shields.io/matrix/rauc:matrix.org?label=matrix%20chat)](https://app.element.io/#/room/#rauc:matrix.org)

The RAUC hawkBit updater is a simple commandline tool / daemon written in C (glib).
It is a port of the RAUC hawkBit Client written in Python.
The daemon runs on your target and operates as an interface between the
[RAUC D-Bus API](https://github.com/rauc/rauc)
and the [hawkBit DDI API](https://github.com/eclipse/hawkbit).

Quickstart
----------

The RAUC hawkBit updater is primarily meant to be used as a daemon,
but it also allows you to do a one shot instantly checking and install
new software.

To quickly getting started with hawkBit server, follow
[this](https://github.com/eclipse/hawkbit#getting-started)
instruction.

Setup target (device) configuration file:

```ini
  [client]
  hawkbit_server            = 127.0.0.1:8080
  ssl                       = false
  ssl_verify                = false
  tenant_id                 = DEFAULT
  target_name               = test-target
  auth_token                = bhVahL1Il1shie2aj2poojeChee6ahShu
  #gateway_token            = bhVahL1Il1shie2aj2poojeChee6ahShu
  #client_cert              = /path/to/client_certificate.pem
  #client_key               = /path/to/client_certificate.key
  bundle_download_location  = /tmp/bundle.raucb
  retry_wait                = 60
  connect_timeout           = 20
  timeout                   = 60
  log_level                 = debug
  post_update_reboot        = false

  [device]
  product                   = Terminator
  model                     = T-1000
  serialnumber              = 8922673153
  hw_revision               = 2
  key1                      = value
  key2                      = value
```

All key/values under [device] group are sent to hawkBit as data (attributes).
The attributes in hawkBit can be used in target filters.

Finally start the updater as daemon:

```shell
  ./rauc-hawkbit-updater -c config.conf
```


Debugging
---------

When setting the log level to 'debug' the RAUC hawkBit client will print
JSON payload sent and received. This can be done by using option -d.

```shell
  ./rauc-hawkbit-updater -d -c config.conf
```


Compile
-------

```shell
  mkdir build
  cd build
  cmake ..
  make
```


Usage / options
---------------

```shell
/usr/bin/rauc-hawkbit-updater --help
Usage:
  rauc-hawkbit-updater [OPTION?]

Help Options:
  -h, --help               Show help options

Application Options:
  -c, --config-file        Configuration file
  -v, --version            Version information
  -d, --debug              Enable debug output
  -r, --run-once           Check and install new software and exit
  -s, --output-systemd     Enable output to systemd
```


CHANGES
Release 1.0 (development)
----------------------------------

.. note::

   This version is the first official release of RAUC hawkBit Updater.

.. rubric:: Enhancements

* Allow following HTTP redirects

.. rubric:: Testing

* Switch from Travis-CI to GitHub actions.
* Add pytest-based testsuite

.. rubric:: Code

* Refactoring

.. rubric:: Documentation

* Add sphinx-based user documentation

.. rubric:: Bug fixes

* Memoryleaks
* Use-after-frees
* Raceconditions
* Fix free storage space calculation overflowing

Release 0.1 (in development)
----------------------------

This is the initial release of RAUC hawkBit Updater.


test/helper.py
import os
import subprocess
import shlex
import logging
import socket
from contextlib import closing


class PExpectLogger:
    """
    pexpect Logger, allows to use Python's logging stdlib. To be passed as pexpect 'logfile".
    Logs linewise to given logger at given level.
    """
    def __init__(self, level=logging.INFO, logger=None):
        self.level = level
        self.data = b''
        self.logger = logger if logger else logging.getLogger()

    def write(self, data):
        self.data += data

    def flush(self):
        for line in self.data.splitlines():
            self.logger.log(self.level, line.decode())

        self.data = b''


def run_pexpect(command, *, timeout=30, cwd=None):
    """
    Runs given command via pexpect with DBUS_STARTER_BUS_TYPE=session and PATH+=./build. Returns
    process handle immediately allowing further expect calls. Logs command (with updated env) and
    its stdout/stderr/exit code.
    """
    import pexpect

    logger = logging.getLogger(command.split()[0])

    env = os.environ.copy()
    env.update({'DBUS_STARTER_BUS_TYPE': 'session'})
    env['PATH'] += f':{os.path.dirname(os.path.abspath(__file__))}/../build'

    log_env = [ f'{key}={value}' for key, value in set(env.items()) - set(os.environ.items()) ]
    logger.info('running: %s %s', ' '.join(log_env), command)

    pexpect_log = PExpectLogger(logger=logger)
    return pexpect.spawn(command, env=env, timeout=timeout, cwd=cwd, logfile=pexpect_log)

def run(command, *, timeout=30):
    """
    Runs given command as subprocess with DBUS_STARTER_BUS_TYPE=session and PATH+=./build. Blocks
    until command terminates. Logs command (with updated env) and its stdout/stderr/exit code.
    Returns tuple (stdout, stderr, exit code).
    """
    env = os.environ.copy()
    env.update({'DBUS_STARTER_BUS_TYPE': 'session'})
    env['PATH'] += f':{os.path.dirname(os.path.abspath(__file__))}/../build'

    logger = logging.getLogger(command.split()[0])

    log_env = [ f'{key}={value}' for key, value in set(env.items()) - set(os.environ.items()) ]
    logger.info('running: %s %s', ' '.join(log_env), command)

    proc = subprocess.run(shlex.split(command), capture_output=True, text=True, check=False,
                          env=env, timeout=timeout)

    for line in proc.stdout.splitlines():
        if line:
            logger.info('stdout: %s', line)
    for line in proc.stderr.splitlines():
        if line:
            logger.warning('stderr: %s', line)

    logger.info('exitcode: %d', proc.returncode)

    return proc.stdout, proc.stderr, proc.returncode

def available_port():
    """Returns an available local port."""
    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
        sock.bind(('localhost', 0))
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        return sock.getsockname()[1]


test/conftest.py
import os
import sys
from configparser import ConfigParser

import pytest

from hawkbit import HawkbitMgmtTestClient, HawkbitError
from helper import run_pexpect, available_port

def pytest_addoption(parser):
    """Register custom argparse-style options."""
    parser.addoption(
        '--hawkbit-instance',
        help='HOST:PORT of hawkBit instance to use (default: %(default)s)',
        default='localhost:8080')

@pytest.fixture(scope='session')
def hawkbit(pytestconfig):
    """Instance of HawkbitMgmtTestClient connecting to a hawkBit instance."""
    from uuid import uuid4

    host, port = pytestconfig.option.hawkbit_instance.split(':')
    client = HawkbitMgmtTestClient(host, int(port))

    client.set_config('pollingTime', '00:00:30')
    client.set_config('pollingOverdueTime', '00:03:00')
    client.set_config('authentication.targettoken.enabled', True)
    client.set_config('authentication.gatewaytoken.enabled', True)
    client.set_config('authentication.gatewaytoken.key', uuid4().hex)

    return client

@pytest.fixture
def hawkbit_target_added(hawkbit):
    """Creates a hawkBit target."""
    target = hawkbit.add_target()
    yield target

    hawkbit.delete_target(target)

@pytest.fixture
def config(tmp_path, hawkbit, hawkbit_target_added):
    """
    Creates a temporary rauc-hawkbit-updater configuration matching the hawkBit (target)
    configuration of the hawkbit and hawkbit_target_added fixtures.
    """
    target = hawkbit.get_target()
    target_token = target.get('securityToken')
    target_name = target.get('name')
    bundle_location = tmp_path / 'bundle.raucb'

    hawkbit_config = ConfigParser()
    hawkbit_config['client'] = {
        'hawkbit_server': f'{hawkbit.host}:{hawkbit.port}',
        'ssl': 'false',
        'ssl_verify': 'false',
        'tenant_id': 'DEFAULT',
        'target_name': target_name,
        'auth_token': target_token,
        'bundle_download_location': str(bundle_location),
        'retry_wait': '60',
        'connect_timeout': '20',
        'timeout': '60',
        'log_level': 'debug',
    }
    hawkbit_config['device'] = {
        'product': 'Terminator',
        'model': 'T-1000',
        'serialnumber': '8922673153',
        'hw_revision': '2',
        'mac_address': 'ff:ff:ff:ff:ff:ff',
    }

    tmp_config = tmp_path / 'rauc-hawkbit-updater.conf'
    with tmp_config.open('w') as f:
        hawkbit_config.write(f)
    return tmp_config

@pytest.fixture
def adjust_config(config):
    """
    Adjusts the rauc-hawkbit-updater configuration created by the config fixture by
    adding/overwriting or removing options.
    """
    config_files = []
    def _adjust_config(options={'client': {}}, remove={}):
        adjusted_config = ConfigParser()
        adjusted_config.read(config)

        # update
        for section, option in options.items():
            for key, value in option.items():
                adjusted_config.set(section, key, value)

        # remove
        for section, option in remove.items():
            adjusted_config.remove_option(section, option)

        with config.open('w') as f:
            adjusted_config.write(f)
        return config

    return _adjust_config

@pytest.fixture(scope='session')
def rauc_bundle(tmp_path_factory):
    """Creates a temporary 512 KB file to be used as a dummy RAUC bundle."""
    bundle = tmp_path_factory.mktemp('data') / 'bundle.raucb'
    bundle.write_bytes(os.urandom(512)*1024)
    return str(bundle)

@pytest.fixture
def bundle_assigned(hawkbit, hawkbit_target_added, rauc_bundle):
    """
    Creates a softwaremodule containing the file from the rauc_bundle fixture as an artifact.
    Creates a distributionset from this softwaremodule. Assigns this distributionset to the target
    created by the hawkbit_target_added fixture. Returns the corresponding action ID of this
    assignment.
    """
    swmodule = hawkbit.add_softwaremodule()
    artifact = hawkbit.add_artifact(rauc_bundle, swmodule)
    distributionset = hawkbit.add_distributionset(swmodule)
    action = hawkbit.assign_target(distributionset)

    yield action

    try:
        hawkbit.cancel_action(action, hawkbit_target_added, force=True)
    except HawkbitError:
        pass

    hawkbit.delete_distributionset(distributionset)
    hawkbit.delete_artifact(artifact, swmodule)
    hawkbit.delete_softwaremodule(swmodule)

@pytest.fixture
def rauc_dbus_install_success(rauc_bundle):
    """
    Creates a RAUC D-Bus dummy interface on the SessionBus mimicing a successful installation on
    Install().
    """
    proc = run_pexpect(f'{sys.executable} -m rauc_dbus_dummy {rauc_bundle}', cwd='test')
    proc.expect('Interface published')

    yield

    assert proc.isalive()
    proc.terminate()

@pytest.fixture
def rauc_dbus_install_failure(rauc_bundle):
    """
    Creates a RAUC D-Bus dummy interface on the SessionBus mimicing a failing installation on
    Install().
    """
    proc = run_pexpect(f'{sys.executable} -m rauc_dbus_dummy {rauc_bundle} --completed-code=1',
                       cwd='test', timeout=None)
    proc.expect('Interface published')

    yield

    assert proc.isalive()
    proc.terminate(force=True)

@pytest.fixture(scope='session')
def nginx_config(tmp_path_factory):
    """
    Creates a temporary nginx proxy configuration incorporating additional given options to the
    location section.
    """
    config_template = """
daemon off;
pid /tmp/hawkbit-nginx-{port}.pid;

# non-fatal alert for /var/log/nginx/error.log will still be shown
# https://trac.nginx.org/nginx/ticket/147
error_log stderr notice;

events {{ }}

http {{
    access_log /dev/null;

    server {{
        listen {port};

        location / {{
            proxy_pass http://localhost:8080;
            {location_options}

            # use proxy URL in JSON responses
            sub_filter "localhost:$proxy_port/" "$host:$server_port/";
            sub_filter "$host:$proxy_port/" "$host:$server_port/";
            sub_filter_types application/json;
            sub_filter_once off;
        }}
    }}
}}"""

    def _nginx_config(port, location_options):
        proxy_config = tmp_path_factory.mktemp('nginx') / 'nginx.conf'
        location_options = ( f'{key} {value};' for key, value in location_options.items())
        proxy_config_str = config_template.format(port=port, location_options=" ".join(location_options))
        proxy_config.write_text(proxy_config_str)
        return proxy_config

    return _nginx_config

@pytest.fixture(scope='session')
def nginx_proxy(nginx_config):
    """
    Runs an nginx rate liming proxy, limiting download speeds to 70 KB/s. HTTP requests are
    forwarded to port 8080 (default port of the docker hawkBit instance). Returns the port the
    proxy is running on. This port can be set in the rauc-hawkbit-updater config to rate limit its
    HTTP requests.
    """
    import pexpect

    procs = []

    def _nginx_proxy(options):
        port = available_port()
        proxy_config = nginx_config(port, options)

        try:
            proc = run_pexpect(f'nginx -c {proxy_config} -p .', timeout=None)
        except (pexpect.exceptions.EOF, pexpect.exceptions.ExceptionPexpect):
            pytest.skip('nginx unavailable')

        try:
            proc.expect('start worker process ')
        except pexpect.exceptions.EOF:
            pytest.skip('nginx failed, use -s to see logs')

        procs.append(proc)

        return port

    yield _nginx_proxy

    for proc in procs:
        assert proc.isalive()
        proc.terminate(force=True)

@pytest.fixture(scope='session')
def rate_limited_port(nginx_proxy):
    """
    Runs an nginx rate liming proxy, limiting download speeds to 70 KB/s. HTTP requests are
    forwarded to port 8080 (default port of the docker hawkBit instance). Returns the port the
    proxy is running on. This port can be set in the rauc-hawkbit-updater config to rate limit its
    HTTP requests.
    """
    def _rate_limited_port(rate):
        location_options = {'proxy_limit_rate': rate}
        return nginx_proxy(location_options)

    return _rate_limited_port


include/config-file.h
/**
 * SPDX-License-Identifier: LGPL-2.1-only
 *
 * Copyright (C) 2018-2020 Prevas A/S (www.prevas.com)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifndef __CONFIG_FILE_H__
#define __CONFIG_FILE_H__

#include <glib.h>

/**
 * @brief struct that contains the Rauc HawkBit configuration.
 */
typedef struct Config_ {
        gchar* hawkbit_server;            /**< hawkBit host or IP and port */
        gboolean ssl;                     /**< use https or http */
        gboolean ssl_verify;              /**< verify https certificate */
        gboolean post_update_reboot;      /**< reboot system after successful update */
        gchar* auth_token;                /**< hawkBit target security token */
        gchar* gateway_token;             /**< hawkBit gateway security token */
        gchar* tenant_id;                 /**< hawkBit tenant id */
        gchar* controller_id;             /**< hawkBit controller id*/
        gchar* bundle_download_location;  /**< file to download rauc bundle to */
        int connect_timeout;              /**< connection timeout */
        int timeout;                      /**< reply timeout */
        int retry_wait;                   /**< wait between retries */
        GLogLevelFlags log_level;         /**< log level */
        GHashTable* device;               /**< Additional attributes sent to hawkBit */
} Config;

/**
 * @brief Get Config for config_file.
 *
 * @param[in]  config_file String value containing path to config file or NULL to search in default
 *                         locations
 * @param[out] error       Error
 * @return Config on success, NULL otherwise (error is set)
 */
Config* load_config_file(const gchar *config_file, GError **error);

/**
 * @brief Frees the memory allocated by a Config
 *
 * @param[in] config Config to free
 */
void config_file_free(Config *config);

G_DEFINE_AUTOPTR_CLEANUP_FUNC(Config, config_file_free)

#endif // __CONFIG_FILE_H__


src/config-file.c
/**
 * SPDX-License-Identifier: LGPL-2.1-only
 *
 * Copyright (C) 2018-2020 Prevas A/S (www.prevas.com)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @file config-file.c
 * @author Lasse Mikkelsen <lkmi@prevas.dk>
 * @date 19 Sep 2018
 * @brief Configuration file parser
 *
 */

#include "config-file.h"
#include <glib/gtypes.h>
#include <stdlib.h>
#include <libeconf.h>

G_DEFINE_AUTOPTR_CLEANUP_FUNC(econf_file, econf_freeFile)


static const gchar *CONFIG_DELIMITER      = "=";
static const gchar *CONFIG_COMMENT        = "#";
static const gchar *CONFIG_EXTENSION      = "conf";
static const gint DEFAULT_CONNECTTIMEOUT  = 20;     // 20 sec.
static const gint DEFAULT_TIMEOUT         = 60;     // 1 min.
static const gint DEFAULT_RETRY_WAIT      = 5 * 60; // 5 min.
static const gboolean DEFAULT_SSL         = TRUE;
static const gboolean DEFAULT_SSL_VERIFY  = TRUE;
static const gboolean DEFAULT_REBOOT      = FALSE;
static const gchar* DEFAULT_LOG_LEVEL     = "message";

/**
 * @brief Get string value from key_file for key in group, optional default_value can be specified
 * that will be used in case key is not found in group.
 *
 * @param[in]  key_file      econf_file to look value up
 * @param[in]  group         A group name
 * @param[in]  key           A key
 * @param[out] value         Output string value
 * @param[in]  default_value String value to return in case no value found, or NULL (not found
 *                           leads to error)
 * @param[out] error         Error
 * @return TRUE if found, TRUE if not found and default_value given, FALSE otherwise (error is set)
 */
static gboolean get_key_string(econf_file *key_file, const gchar *group, const gchar *key,
                               gchar **value, const gchar *default_value, GError **error)
{
        g_autofree gchar *default_nonconst = g_strdup(default_value);
        econf_err econf_error;

        g_return_val_if_fail(key_file, FALSE);
        g_return_val_if_fail(group, FALSE);
        g_return_val_if_fail(key, FALSE);
        g_return_val_if_fail(value && *value == NULL, FALSE);
        g_return_val_if_fail(error == NULL || *error == NULL, FALSE);

        if (default_value)
                econf_error = econf_getStringValueDef(key_file, group, key, value,
                                                      default_nonconst);
        else
                econf_error = econf_getStringValue(key_file, group, key, value);

        if (!econf_error)
                return TRUE;
        if (default_value && econf_error == ECONF_NOKEY)
                return TRUE;

        g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_PARSE,
                    "Key '%s' (string, group '%s'): %s", key, group, econf_errString(econf_error));
        return FALSE;
}

/**
 * @brief Get gboolean value from key_file for key in group, default_value must be specified,
 * returned in case key not found in group.
 *
 * @param[in]  key_file      econf_file to look value up
 * @param[in]  group         A group name
 * @param[in]  key           A key
 * @param[out] value         Output gboolean value
 * @param[in]  default_value Return this value in case no value found
 * @param[out] error         Error
 * @return FALSE on error (error is set), TRUE otherwise. Note that TRUE is returned if key in
 *         group is not found, value is set to default_value in this case.
 */
static gboolean get_key_bool(econf_file *key_file, const gchar *group, const gchar *key,
                             gboolean *value, const gboolean default_value, GError **error)
{
        econf_err econf_error;

        g_return_val_if_fail(key_file, FALSE);
        g_return_val_if_fail(group, FALSE);
        g_return_val_if_fail(key, FALSE);
        g_return_val_if_fail(value, FALSE);
        g_return_val_if_fail(error == NULL || *error == NULL, FALSE);

        econf_error = econf_getBoolValueDef(key_file, group, key, (bool *)value, default_value);
        if (econf_error && econf_error != ECONF_NOKEY) {
                g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_PARSE,
                            "Key '%s' (bool, group '%s'): %s", key, group,
                            econf_errString(econf_error));
                return FALSE;
        }

        return TRUE;
}

/**
 * @brief Get integer value from key_file for key in group, default_value must be specified,
 * returned in case key not found in group.
 *
 * @param[in]  key_file      econf_file to look value up
 * @param[in]  group         A group name
 * @param[in]  key           A key
 * @param[out] value         Output integer value
 * @param[in]  default_value Return this value in case no value found
 * @param[out] error         Error
 * @return FALSE on error (error is set), TRUE otherwise. Note that TRUE is returned if key in
 *         group is not found, value is set to default_value in this case.
 */
static gboolean get_key_int(econf_file *key_file, const gchar *group, const gchar *key,
                            gint *value, const gint default_value, GError **error)
{
        econf_err econf_error;

        g_return_val_if_fail(key_file, FALSE);
        g_return_val_if_fail(group, FALSE);
        g_return_val_if_fail(key, FALSE);
        g_return_val_if_fail(value, FALSE);
        g_return_val_if_fail(error == NULL || *error == NULL, FALSE);

        econf_error = econf_getIntValueDef(key_file, group, key, value, default_value);
        if (econf_error && econf_error != ECONF_NOKEY) {
                g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_PARSE,
                            "Key '%s' (int, group '%s'): %s", key, group,
                            econf_errString(econf_error));
                return FALSE;
        }

        return TRUE;
}

/**
 * @brief Get GHashTable containing keys/values from group in key_file.
 *
 * @param[in]  key_file econf_file to look value up
 * @param[in]  group    A group name
 * @param[out] hash     Output GHashTable
 * @param[out] error    Error
 * @return TRUE on keys/values stored successfully, FALSE on empty group/value or on other errors
 *         (error set)
 */
static gboolean get_group(econf_file *key_file, const gchar *group, GHashTable **hash,
                          GError **error)
{
        g_autoptr(GHashTable) tmp_hash = NULL;
        size_t key, num_keys = 0;
        g_auto(GStrv) keys = NULL;
        econf_err econf_error;

        g_return_val_if_fail(key_file, FALSE);
        g_return_val_if_fail(group, FALSE);
        g_return_val_if_fail(hash && *hash == NULL, FALSE);
        g_return_val_if_fail(error == NULL || *error == NULL, FALSE);

        tmp_hash = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);

        econf_error = econf_getKeys(key_file, group, &num_keys, &keys);
        if (econf_error) {
                g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_PARSE,
                            "Enumerating keys in group '%s' failed: %s", group,
                            econf_errString(econf_error));
                return FALSE;
        }

        for (key = 0; key < num_keys; key++) {
                gchar *value = NULL;
                if (!get_key_string(key_file, group, keys[key], &value, NULL, error))
                        return FALSE;

                g_hash_table_insert(tmp_hash, g_strdup(keys[key]), value);
        }

        *hash = g_steal_pointer(&tmp_hash);
        return TRUE;
}

/**
 * @brief Get GLogLevelFlags for error string.
 *
 * @param[in]  log_level Log level string
 * @return GLogLevelFlags matching error string, else default log level (message)
 */
static GLogLevelFlags log_level_from_string(const gchar *log_level)
{
        g_return_val_if_fail(log_level, 0);

        if (g_strcmp0(log_level, "error") == 0) {
                return G_LOG_LEVEL_ERROR;
        } else if (g_strcmp0(log_level, "critical") == 0) {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL;
        } else if (g_strcmp0(log_level, "warning") == 0) {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL |
                       G_LOG_LEVEL_WARNING;
        } else if (g_strcmp0(log_level, "message") == 0) {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL |
                       G_LOG_LEVEL_WARNING | G_LOG_LEVEL_MESSAGE;
        } else if (g_strcmp0(log_level, "info") == 0) {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL |
                       G_LOG_LEVEL_WARNING | G_LOG_LEVEL_MESSAGE |
                       G_LOG_LEVEL_INFO;
        } else if (g_strcmp0(log_level, "debug") == 0) {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL |
                       G_LOG_LEVEL_WARNING | G_LOG_LEVEL_MESSAGE |
                       G_LOG_LEVEL_INFO | G_LOG_LEVEL_DEBUG;
        } else {
                g_warning("Invalid log level given, defaulting to level \"message\"");
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL |
                       G_LOG_LEVEL_WARNING | G_LOG_LEVEL_MESSAGE;
        }
}

static gboolean parse_default_config_locations(econf_file **key_file, GError **error)
{
        g_autoptr(econf_file) etc_run_config = NULL;
        g_autoptr(econf_file) usr_config = NULL;
        econf_err econf_error;

        g_return_val_if_fail(key_file && *key_file == NULL, FALSE);
        g_return_val_if_fail(error == NULL || *error == NULL, FALSE);

        // files /etc/ take precedence over those in /run/
        econf_error = econf_readDirs(&etc_run_config, "/run/rauc-hawkbit-updater",
                                     "/etc/rauc-hawkbit-updater", "rauc-hawkbit-updater",
                                     CONFIG_EXTENSION, CONFIG_DELIMITER, CONFIG_COMMENT);
        if (econf_error && econf_error != ECONF_NOFILE) {
                g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_PARSE,
                            "Failed reading config locations /run, /etc: %s",
                            econf_errString(econf_error));
                return FALSE;
        }

        // consider files in /usr/lib/
        econf_error = econf_readDirs(&usr_config, "/usr/lib/rauc-hawkbit-updater", NULL,
                                     "rauc-hawkbit-updater", CONFIG_EXTENSION, CONFIG_DELIMITER,
                                     CONFIG_COMMENT);
        if (econf_error && econf_error != ECONF_NOFILE) {
                g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_PARSE,
                            "Failed reading config location /usr: %s",
                            econf_errString(econf_error));
                return FALSE;
        }

        // consider valid key files
        if (!etc_run_config && !usr_config) {
                g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_PARSE, "No config files found");
                return FALSE;
        }
        if (etc_run_config && !usr_config)
                *key_file = g_steal_pointer(&etc_run_config);
        if (!etc_run_config && usr_config)
                *key_file = g_steal_pointer(&usr_config);
        if (etc_run_config && usr_config) {
                // files in /etc/ and /run/ (parsed above) take precedence over those in /usr/lib/
                econf_error = econf_mergeFiles(key_file, usr_config, etc_run_config);
                if (econf_error) {
                        g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_PARSE,
                                    "Failed merging config locations: %s",
                                    econf_errString(econf_error));
                        return FALSE;
                }
        }

        return TRUE;
}

Config* load_config_file(const gchar *config_file, GError **error)
{
        g_autoptr(Config) config = NULL;
        g_autofree gchar *val = NULL;
        g_autoptr(econf_file) ini_file = NULL;
        gboolean key_auth_token_exists = FALSE;
        gboolean key_gateway_token_exists = FALSE;
        econf_err econf_error;

        g_return_val_if_fail(error == NULL || *error == NULL, NULL);

        config = g_new0(Config, 1);

        if (config_file) {
                econf_error = econf_readFile(&ini_file, config_file, CONFIG_DELIMITER,
                                             CONFIG_COMMENT);
                if (econf_error) {
                        g_set_error_literal(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_PARSE,
                                            econf_errString(econf_error));
                        return NULL;
                }
        } else {
                if (!parse_default_config_locations(&ini_file, error))
                        return NULL;
        }

        if (!get_key_string(ini_file, "client", "hawkbit_server", &config->hawkbit_server, NULL,
                            error))
                return NULL;

        key_auth_token_exists = get_key_string(ini_file, "client", "auth_token",
                                               &config->auth_token, NULL, NULL);
        key_gateway_token_exists = get_key_string(ini_file, "client", "gateway_token",
                                                  &config->gateway_token, NULL, NULL);
        if (!key_auth_token_exists && !key_gateway_token_exists) {
                g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_INVALID_VALUE,
                            "Neither auth_token nor gateway_token is set in the config.");
                return NULL;
        }
        if (key_auth_token_exists && key_gateway_token_exists) {
                g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_INVALID_VALUE,
                            "Both auth_token and gateway_token are set in the config.");
                return NULL;
        }

        if (!get_key_string(ini_file, "client", "target_name", &config->controller_id, NULL,
                            error))
                return NULL;
        if (!get_key_string(ini_file, "client", "tenant_id", &config->tenant_id, "DEFAULT", error))
                return NULL;
        if (!get_key_string(ini_file, "client", "bundle_download_location",
                            &config->bundle_download_location, NULL, error))
                return NULL;
        if (!get_key_bool(ini_file, "client", "ssl", &config->ssl, DEFAULT_SSL, error))
                return NULL;
        if (!get_key_bool(ini_file, "client", "ssl_verify", &config->ssl_verify,
                          DEFAULT_SSL_VERIFY, error))
                return NULL;
        if (!get_key_int(ini_file, "client", "connect_timeout", &config->connect_timeout,
                         DEFAULT_CONNECTTIMEOUT, error))
                return NULL;
        if (!get_key_int(ini_file, "client", "timeout", &config->timeout, DEFAULT_TIMEOUT, error))
                return NULL;
        if (!get_key_int(ini_file, "client", "retry_wait", &config->retry_wait, DEFAULT_RETRY_WAIT,
                         error))
                return NULL;
        if (!get_key_string(ini_file, "client", "log_level", &val, DEFAULT_LOG_LEVEL, error))
                return NULL;
        config->log_level = log_level_from_string(val);

        if (!get_key_bool(ini_file, "client", "post_update_reboot", &config->post_update_reboot, DEFAULT_REBOOT, error))
                return NULL;

        if (config->timeout > 0 && config->connect_timeout > 0 &&
            config->timeout < config->connect_timeout) {
                g_set_error(error,
                            G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_INVALID_VALUE,
                            "timeout (%d) must be greater than connect_timeout (%d)",
                            config->timeout, config->connect_timeout);
                return NULL;
        }

        if (!get_group(ini_file, "device", &config->device, error))
                return NULL;

        return g_steal_pointer(&config);
}

void config_file_free(Config *config)
{
        if (!config)
                return;

        g_free(config->hawkbit_server);
        g_free(config->controller_id);
        g_free(config->tenant_id);
        g_free(config->auth_token);
        g_free(config->gateway_token);
        g_free(config->bundle_download_location);
        if (config->device)
                g_hash_table_destroy(config->device);
        g_free(config);
}
